### ISSUE TITLE
Generate code for named struct patterns

### ISSUE THREAD

**Issue by fabianschuiki** [fabianschuiki/moore#103]
Generate code for named struct patterns such as:
```
typedef struct packed {
    tag_t       tag;
    logic       sign_ext;
    logic [2:0] offset;
    logic [1:0] size;
} laq_t;

laq_t laq_in;

assign laq_in = '{
    tag:      lsu_qtag_i,
    sign_ext: lsu_qsigned,
    offset:   lsu_qaddr_i[2:0],
    size:     lsu_qsize_i
};
```
This requires being able to infer the type of the pattern expression since it is used in an "assignment-like context". Requires some work in typeck.

**Comment by fabianschuiki** [comment]
Rough path to solution:

1. Determine the type the pattern will have from the context. This must be possible as per SV standard, otherwise the input is malformed.
2. If the destination type is an array, treat each non-type key as a constant expression indicating the position in the array. If it is a struct, treat each non-type key as a member name and resolve the respective position.
3. Add a query to the compiler that returns a fully resolved pattern assignment which clearly associates each field/element in the destination type with an expression.
4. It might be necessary to generate an expression MIR (mid-level intermediate representation), since one pattern can imply a lot of implicit casting.

**Comment by fabianschuiki** [comment]
Has been implemented in fa6fd16.
